PEP 725: Add details about `vers` ranges and `abstract/` types

jaimergp opened the pull request:
<!-- readthedocs-preview pep-previews start -->
----
📚 Documentation preview 📚: https://pep-previews--21.org.readthedocs.build/

<!-- readthedocs-preview pep-previews end -->

mgorny commented on peps/pep-0725.rst:
```diff
@@ -217,15 +223,52 @@ There is no ready-made support for virtual packages in PURL or another
+
+- The scheme for this PURL derivative is ``dep:`` (for dependency).
+- When no operator is present, the field is understood as a fixed version.
+- The version range syntax must be compatible with PEP 440. As a result, it is
+  equivalent to ``vers:pypi/`` PURLs.
```
The problem with this is that it won't be able to express some versions, particularly OpenSSL versions (that use letters `a`, `b`, … as post-releases, rather than prereleases.

jaimergp replied:
Yep, same with libjpeg IIRC. In those cases I think we need to fallback to `[pkg, vers]` tuples. Thankfully versions like `1.1.1a` are illegal in PEP440, so at least there's no ambiguity there.

rgommers replied:
I think I'm fine with declaring that "out of scope - corner case that can't be expressed" if we don't want the extra complexity of really allowing `[pkg, vers]` tuples. Can still mostly express it with `>1.1.1,<1.1.2.dev0` anyway.

rgommers commented on peps/pep-0725.rst:
```diff
@@ -169,6 +173,8 @@ necessary to run a host dependency under an emulator, or through a custom tool
 like crossenv_. When host dependencies imply a runtime dependency, that runtime
 dependency also does not have to be declared, just like for build dependencies.
 
+.. JRG: Why? are host deps it added implicitly as runtime ones?
```
Ah, you caught confused terminology. "host dependencies" is defined higher up as _"build-time dependencies built for the host machine"_. That could use some cleaning up indeed.

rgommers commented on peps/pep-0725.rst:
```diff
@@ -217,15 +223,52 @@ There is no ready-made support for virtual packages in PURL or another
+However, this is not very ergonomic. We propose a PURL-compatible syntax that
+extends the ``@`` component to support version ranges directly:
+
+.. JRG: I suggested 'dep:' (dependency) but it could also be 'req:' (requirement)
```
I like `dep:`.

rgommers commented on peps/pep-0725.rst:
```diff
@@ -414,7 +457,7 @@ SciPy 1.10:
     host-requires = [
       "virtual:interface/blas",
-      "virtual:interface/lapack",  # >=3.7.1 (can't express version ranges with PURL yet)
+      "dep:interface/lapack@>=3.7.1",
```
After this change there's a mix of `virtual:` and `dep:interface` in the document. Why would we allow both?

rgommers replied:
> I really prefer having a single scheme.

I think you are right here. `dep:interface` seems fine to me, and nicer than a separate `virtual:`.

jaimergp replied:
Fixed this now.

rgommers reviewed (commented):
Thanks Jaime, and sorry for the delayed review. Overall this LGTM. Here are some comments & thoughts.

jaimergp commented:
@rgommers I addressed your comments, and also adjusted the whole text so it refers to `dep:` URLs primarily. I tried my best to define them as PURL derivatives and introduce their particularities in the existing sections. Consider this a first pass as we polish it. Once we like it, I will send PRs to the relevant example repositories (your `py-show` tool, the Streamlit mappings).

rgommers commented:
Thanks Jaime. Having a closer look at this after the `abstract` -> `virtual` name revert that we just discussed.

rgommers commented on peps/pep-0725.rst:
```diff
@@ -203,29 +212,73 @@ and the `Sonatype OSS Index <https://ossindex.sonatype.org/doc/coordinates>`__;
+distros with virtual packages and to the likes of conda and Spack.
+
+.. JRG: A long shot for now. Assuming virtual PURLs would be formatted
+.. like ``pkg:virtual/{compiler,interface}/...``.
```
We have to clean up comments before submitting the PEP revision upstream. I assume the PURL conversation won't move before then.

rgommers commented on peps/pep-0725.rst:
```diff
@@ -252,11 +305,40 @@ metadata fields. The ``optional-dependencies`` content from ``[external]``
+
+- The ``dep`` *scheme* MUST be replaced with ``pkg``.
+- If the *type* is ``compiler`` or ``interface``, it MUST be transformed
+  into a ``virtual/(compiler|interface)`` type.
```
That doesn't exist? I'm not sure how this can be handled. The section above says we're omitting a few things, so that may apply here too maybe. That approach may change anyway. Given the recent core metadata bumping, there may be a request to do so for this PEP (I'd prefer not to though, I successfully argued for SBOMs not to need this).

Let's discuss this one.

jaimergp replied:
We can also abuse the `Requires-External` metadata and do something hacky like `; False and ...` 😬, but that feels so wrong.

jaimergp replied:
The other solution is to drop `virtual` stuff from here, but then their value is diminished and folks might be tempted to write those as `dep:generic/gcc` instead of `dep:virtual/compiler/c`.

rgommers replied:
> 😬, but that feels so wrong.

agreed

> The other solution is to drop `virtual` stuff from here

I'd much prefer to keep it as you have it now.

rgommers commented on peps/pep-0725.rst:
```diff
@@ -252,11 +305,40 @@ metadata fields. The ``optional-dependencies`` content from ``[external]``
+    value, followed by a forward slash. If the *type* does not have a
+    matching ``vers`` type, ``pypi`` MUST be used.
+
+- The PURL percent-encoding rules MUST be applied.
```
Those rules are pretty terrible. Wouldn't we be better off with `(ppg, vers)` tuples if it's autogenerated anyway?

jaimergp replied:
Turns out that `Requires-External` [does have some syntax rules](https://packaging.python.org/en/latest/specifications/core-metadata/#requires-external-multiple-use) (`name (version); qualifier`), so I rephrased this section so it complies with that loose format.

rgommers replied:
Ah, that explains it, good find!

I think we have to keep it like that then. Or if we want to change it, we must do a core metadata version bump and add new fields for everything.

jaimergp replied:
Yes, I kind of like it, even. The only problem is that we can't use any of `();` in PURLs, but I cannot think of examples that require those right now, and if it comes up, %-encoding to the rescue. The packaging docs do not cover that case either, so let's leave it to parsers to figure it our (split on ; from the right, regex on () on the left side, etc).

rgommers reviewed (commented):
Pretty close, thanks @jaimergp. A couple of questions inline about version ranges and core metadata.

rgommers commented on peps/pep-0725.rst:
```diff
@@ -101,14 +102,14 @@ Multiple types of external dependencies can be distinguished:
 -----------------
 
-Cross compilation is not yet (as of August 2023) well-supported by stdlib
+Cross compilation is not yet (as of March 2025) well-supported by stdlib
```
might as well make this May now

jaimergp replied:
```suggestion
Cross compilation is not yet (as of May 2025) well-supported by stdlib
```

jaimergp commented on peps/pep-0725.rst:
```diff
@@ -252,11 +303,37 @@ metadata fields. The ``optional-dependencies`` content from ``[external]``
+  next to it, between parentheses and after a space, with the following
+  transformations:
+
+  - If the value is a literal (it contains no operators), it MUST be kept as is.
```
Done!

rgommers commented:
@pradyunsg we're trying to work out the remaining open issues, one or two more iterations should be coming. I heard this PEP was also discussed at the packaging summit; will reach out to you soon about next steps.
